### PR TITLE
introduce ovirt set mac pool role

### PR DESCRIPTION
**Pull request by tareqalayan**

### CODE REVIEW

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by mwperina** [comment]
ci test please

File: roles/ovirt-infra/README.md [hunk 1]
```diff
@@ -21,7 +21,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
  - `data_center_description`
  - `data_center_local`
  - `compatibility_version`
- 
+
+* [ovirt-set-mac-pools]:
+
+ This role expects a variable called `mac-ranges` and a variable called `mac_pool_allow_duplicates`
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Please be consitent with other descriptions:

```
This role accpts a variable called ....
```

File: roles/ovirt-infra/README.md [hunk 2]
```diff
@@ -21,7 +21,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
  - `data_center_description`
  - `data_center_local`
  - `compatibility_version`
- 
+
+* [ovirt-set-mac-pools]:
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Can you please remove `set` from the name?

File: roles/ovirt-set-mac-pools/.travis.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/.travis.yml]: Please remove this file.

File: roles/ovirt-set-mac-pools/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
+oVirt Mac pools
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/README.md]: oVirt Mac Pools

File: roles/ovirt-infra/README.md [hunk 5]
```diff
@@ -80,7 +85,10 @@ Example Playbook
      
      data_center_name: mydatacenter
      compatibility_version: 4.1
-     
+
+     mac_pool_allow_duplicates: false
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: In documentation you use different names.

File: roles/ovirt-set-mac-pools/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+----------------
+
+```yaml
+- name: oVirt set mac pool
+  hosts: localhost
+  connection: local
+  gather_facts: false
+
+  vars:
+   mac_pool_allow_duplicates: false
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/README.md]: In doc, there's different names.

File: roles/ovirt-set-mac-pools/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/defaults/main.yml]: please remove this file

File: roles/ovirt-set-mac-pools/handlers/main.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+---
+# handlers file for roles/ovirt-set-mac-pools
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/handlers/main.yml]: please remove this file

File: roles/ovirt-set-mac-pools/tests/test.yml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/tests/test.yml]: please remove this file

File: roles/ovirt-set-mac-pools/tests/inventory [hunk 10]
```diff
@@ -0,0 +1 @@
+localhost
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/tests/inventory]: please remove this file

File: roles/ovirt-set-mac-pools/vars/main.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/vars/main.yml]: please remove this file

File: roles/ovirt-set-mac-pools/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+---
+- name: set mac pool
+  ovirt_mac_pools:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ mac_pool_name | default('Default') }}"
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/tasks/main.yml]: So this by default always create 'Default' mac pool, why?

File: roles/ovirt-set-mac-pools/README.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+ * oVirt Python SDK version 4
+ * Ansible version 2.3
+
+Role Variables
+--------------
+
+| Name                     | Default value         | Description                          |
+|--------------------------|-----------------------|--------------------------------------|
+| allow_duplicates         | UNDEF                 | If (true) allow a MAC address to be used multiple times in a pool.
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/README.md]: I think default is false in your role, not by ovirt engine, so UNDEF is not true also

File: roles/ovirt-set-mac-pools/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+    allow_duplicates: "{{ mac_pool_allow_duplicates | default(false) }}"
+    ranges:
+      - "{{ mac_ranges }}"
```
Comment (machacekondra) [inline roles/ovirt-set-mac-pools/tasks/main.yml]: default(omit) would be resonable here

**Review by machacekondra — CHANGES_REQUESTED** [review]
I think we should also attach the mac pool to DC or cluster, otherwise this role doesn't make much sense I think

File: roles/ovirt-infra/README.md (hunk 2 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-infra/README.md (hunk 1 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-set-mac-pools/.travis.yml (hunk 3 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/.travis.yml]: done

File: roles/ovirt-set-mac-pools/tasks/main.yml (hunk 14 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/tasks/main.yml]: done

File: roles/ovirt-set-mac-pools/tasks/main.yml (hunk 12 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/tasks/main.yml]: This is the default one comes with the fresh installed ovirt.

File: roles/ovirt-set-mac-pools/tests/inventory (hunk 10 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/tests/inventory]: done

File: roles/ovirt-set-mac-pools/tests/test.yml (hunk 9 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/tests/test.yml]: done

File: roles/ovirt-set-mac-pools/vars/main.yml (hunk 11 above)
Comment (tareqalayan) [reply roles/ovirt-set-mac-pools/vars/main.yml]: done

**Review by tareqalayan — COMMENTED** [review]
done

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

File: roles/ovirt-infra/README.md (hunk 2 above)
Comment (machacekondra) [reply roles/ovirt-infra/README.md]: Can you please remove set from the name?

File: roles/ovirt-infra/README.md [hunk 15]
```diff
@@ -21,7 +21,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
  - `data_center_description`
  - `data_center_local`
  - `compatibility_version`
- 
+
+* [ovirt-set-mac-pools]:
+
+ This role accepts a variable called `mac-ranges` and a variable called `mac_pool_allow_duplicates`
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: s/mac-ranges/mac_ranges

Comment (machacekondra) [reply roles/ovirt-infra/README.md]: What about `mac_name`

File: roles/ovirt-infra/README.md [hunk 16]
```diff
@@ -61,6 +65,7 @@ Dependencies
  * [ovirt-storages]
  * [ovirt-aaa-jdbc]
  * [ovirt-permissions]
+ * [ovirt-set-mac-pools]
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Can you please remove set from the name?

File: roles/ovirt-infra/README.md [hunk 17]
```diff
@@ -216,3 +224,4 @@ Apache License 2.0
 [ovirt-networks]: https://github.com/oVirt/ovirt-ansible/blob/master/roles/ovirt-networks/README.md
 [ovirt-permissions]: https://github.com/oVirt/ovirt-ansible/blob/master/roles/ovirt-permissions/README.md
 [ovirt-storages]: https://github.com/oVirt/ovirt-ansible/blob/master/roles/ovirt-storages/README.md
+[ovirt-set-mac-pools]: https://github.com/oVirt/ovirt-ansible/blob/master/roles/ovirt-set-mac-pools/README.md
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Can you please remove set from the name?

File: roles/ovirt-mac-pools/defaults/main.yml [hunk 18]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/defaults/main.yml]: Can you please remove this file?

File: roles/ovirt-mac-pools/handlers/main.yml [hunk 19]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/handlers/main.yml]: Can you please remove this file?

File: roles/ovirt-mac-pools/README.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  hosts: localhost
+  connection: local
+  gather_facts: false
+
+  vars:
+   mac_pool_allow_duplicates: false
+   mac_ranges: 00:1a:4a:16:01:51,00:1a:4a:16:01:61
+
+  roles:
+    - ovirt-set-mac-pools
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/README.md]: Can you please remove set from the name?

File: roles/ovirt-mac-pools/README.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+ * oVirt Python SDK version 4
+ * Ansible version 2.3
+
+Role Variables
+--------------
+
+| Name                      | Default value         | Description                                                       |
+|---------------------------|-----------------------|-------------------------------------------------------------------|
+| mac_pool_allow_duplicates | UNDEF                 | If (true) allow a MAC address to be used multiple times in a pool.
+                                                      Default value is set by oVirt engine to false.                    |
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/README.md]: This is improperly formatted, please verify the formatting.

File: roles/ovirt-mac-pools/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,10 @@
+---
+- name: set mac pool
+  ovirt_mac_pools:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ mac_pool_name | default('Default') }}"
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/tasks/main.yml]: Do we want to always by default create a 'Default' mac pool range? I don't think it's good idea.

**Review by machacekondra — CHANGES_REQUESTED** [review]
Again, we need to attach the mac-pool to cluster or datacenter, otherwise it doesn't make sense.

File: roles/ovirt-infra/README.md (hunk 2 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-infra/README.md (hunk 15 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-infra/README.md (hunk 16 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-infra/README.md (hunk 5 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: you mean ovirt-mac-pool/README.md ? 
there it is also mac_pool_allow_duplicates

File: roles/ovirt-infra/README.md (hunk 17 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-mac-pools/README.md (hunk 21 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/README.md]: done

File: roles/ovirt-mac-pools/README.md (hunk 20 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/README.md]: done

File: roles/ovirt-mac-pools/defaults/main.yml (hunk 18 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/defaults/main.yml]: done

File: roles/ovirt-mac-pools/handlers/main.yml (hunk 19 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/handlers/main.yml]: done

File: roles/ovirt-mac-pools/tasks/main.yml (hunk 22 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/tasks/main.yml]: it is the mac_pool found on the engine, i guess this is the one we wwant to set if mac_pool_name was not specified. 

what do u suggest?

Comment (machacekondra) [reply roles/ovirt-mac-pools/tasks/main.yml]: in current flow if I understand correctly, if user don't specify anything. the `ovirt-infra` role will try to remove 'ranges' from 'Default' mac-pool, I think it's not what you want, or is it?

Comment (machacekondra) [reply roles/ovirt-mac-pools/tasks/main.yml]: No, actually it fails, that `mac_ranges` is not defined ... it shouldn't

Comment (machacekondra) [reply roles/ovirt-mac-pools/tasks/main.yml]: Also if possible please rename `mac_ranges` to `mac_pool_ranges` so we are consistent with mac_pool_name and mac_pool_allow_duplicates

Comment (tareqalayan) [reply roles/ovirt-mac-pools/tasks/main.yml]: OK.
so mac_pool_ranges and mac_pool_name should be provided by user.

**Comment by machacekondra** [comment]
Will we need in future more Mac pool ranges? If yes, we should have it as a list.

File: roles/ovirt-mac-pools/tasks/main.yml [hunk 23]
```diff
@@ -0,0 +1,13 @@
...
+- name: set mac pool
+  ovirt_mac_pools:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ mac_pool_name }}"
+    allow_duplicates: "{{ mac_pool_allow_duplicates | default(omit) }}"
+    ranges:
+      - "{{ mac_pool_ranges }}"
+  when:
+    - mac_pool_ranges is defined
+    - mac_pool_name is define
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/tasks/main.yml]: s/define/defined Please verify

Comment (tareqalayan) [reply roles/ovirt-mac-pools/tasks/main.yml]: fixed. 
Maybe we will need a list .. i will try to fix it.

**Review by tareqalayan — COMMENTED** [review]
mac_ranges can be list as well.

**Comment by machacekondra** [comment]
And what about multiple mac pools?

**Comment by tareqalayan** [comment]
support list of mac_pools

**Comment by tareqalayan** [comment]
@machacekondra 
i am looking into http://docs.ansible.com/ansible/latest/ovirt_clusters_module.html and don't see any mac_pool option. 
i think in 4.1 and 4.2 the mac_pool should be attached to cluster level. 
if the name of the mac_pool is Default it will be attached by default to the cluster. This is what we do. 
but i agree we should attach it explicitly to cluster. 
but i cannot find it in documentation.

**Comment by tareqalayan** [comment]
@machacekondra  i opened issue https://github.com/ansible/ansible/issues/27943 about ovirt_cluster not supporting the mac_pool, can u please assign it to me?
how do you want to procceed with this PR?

**Comment by machacekondra** [comment]
> mac_pool, can u please assign it to me?

I don't have permissions to do assignments in upstream Ansible.

> how do you want to procceed with this PR?

I think we can merge it and we do additional PR for clusters roles to assign mac pool.

File: roles/ovirt-mac-pools/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+- name: set mac pools
+  ovirt_mac_pools:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ item.mac_pool_name }}"
+    allow_duplicates: "{{ item.mac_pool_allow_duplicates | default(omit) }}"
+    ranges: "{{ item.mac_pool_ranges }}"
+  when:
+    - mac_pools is defined
+  with_items:
+    - "{{ mac_pools }}"
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/tasks/main.yml]:
```yaml
with_items:
 - "{{ mac_pools | default([]) }}"
```
and remove `when`

File: roles/ovirt-infra/README.md [hunk 25]
```diff
@@ -21,7 +21,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
  - `data_center_description`
  - `data_center_local`
  - `compatibility_version`
- 
+
+* [ovirt-mac-pools]:
+
+ This role accepts a variable called `mac_pool_ranges`, `mac_pool_allow_duplicates` and `mac_pool_name`.
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Now it accept `mac_pools`

File: roles/ovirt-infra/README.md [hunk 26]
```diff
@@ -80,7 +85,12 @@ Example Playbook
      
      data_center_name: mydatacenter
      compatibility_version: 4.1
-     
+
+     mac_pool_name: my_mac_pool
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Please change to `mac_pools:`

File: roles/ovirt-mac-pools/README.md [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+ * Ansible version 2.3
+
+Role Variables
+--------------
+
+| Name                  | Default value         |  Description                                              |
+|-----------------------|-----------------------|-----------------------------------------------------------|
+| mac_pools             | UNDEF                 | List of dictionaries that describe the mac pool.          |
+
+
```
Comment (machacekondra) [inline roles/ovirt-mac-pools/README.md]: Please add here : 
```
The items in `mac_pools` list can contain the following parameters:
```

File: roles/ovirt-infra/README.md (hunk 25 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-infra/README.md (hunk 26 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-mac-pools/README.md (hunk 27 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/README.md]: done

File: roles/ovirt-mac-pools/tasks/main.yml (hunk 24 above)
Comment (tareqalayan) [reply roles/ovirt-mac-pools/tasks/main.yml]: done

File: roles/ovirt-infra/README.md [hunk 28]
```diff
@@ -21,7 +21,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
  - `data_center_description`
  - `data_center_local`
  - `compatibility_version`
- 
+
+* [ovirt-mac-pools]:
+
+ This role accepts a variable called `mac_pools.
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: missing '`' at the end

**Review by machacekondra — APPROVED** [review]
Looks good, one last nitpick. Please fix and I will merge.

**Comment by machacekondra** [comment]
ci build please

**Pull request merged**